model/licensing.py

```diff
@@ -1368,6 +1374,17 @@ def delete(self):
         """Delete a LicensePoolDeliveryMechanism."""
         _db = Session.object_session(self)
         pools = list(self.license_pools)
+
+        # Except deleting this record itself, we also need to remove all linked Hyperlink records.
+        if self.resource_id:
+            from resource import Resource
+            resource = get_one(_db, Resource, id=self.resource_id)
+
+            for hyperlink in resource.links:
+                _db.delete(hyperlink)
+
+            _db.delete(resource)
+
```

Comment:
This doesn't seem right to me. Two different LicensePoolDeliveryMechanisms are allowed to have the same Resource. This is unusual but it could happen if two different OPDS feeds linked to the same open-access book at the same URL.

Can you talk about the problem you're seeing here in the situation when a single OPDS feed changes which Resource it's linking to for a given book?

Reply:
Judging from your earlier comments, I feel like this is code that should be running in `Metadata.apply`, near this comment:

```
# Associate all links with the primary identifier.
```

That code should delete any Hyperlinks that used to be associated with this book's identifier but no longer are. And it shouldn't be necessary to delete the Resource itself (since another book might be using it). Why isn't that code working for you?

Reply:
I think it's related to the fact that `LicensePool.best_open_access_link` relies on `LicensePool.best_open_access_resource` which in turn relies on `LicensePool.open_access_links` which fetches all resources associated with the pool's identifier which is probably not right.
I'm using the dump of the production system where we have multiple resources from different data sources associated with the same identifier.
It means that the code `Metadata.apply` remains hyperlinks from different data sources untouched, removing only hyperlinks from the current data source. Later, after we reset `LicensePool._open_access_download_url` by calling `reset_open_access_download_url` method, it tries to refresh the acquisition link when `LicensePool.best_open_access_link` gets called during the feed generation. Considering the fact that `LicensePool.open_access_links` fetches all the resources it ends up with wrong not updated resources belonging to different resources and shows the old acquisition link.
The other option here would be to explicitly specify the data source when calling `Identifier.resources_for_identifier_ids` in `LicensePool.open_access_links`. I'll test this approach.

Reply:
@leonardr, the change I did (got rid of removing resources in `LicensePoolDeliveryMechanism.delete` and refined filtration of resources by the data source in `LicensePool.open_access_links`) surprisingly worked in the local environment I recreated from the production dump but it caused `TestOPDSImporter.test_importer_updates_acquisition_and_cover_links` test to fail.
The problem is that after the second import we end up with two resources: old one and a new one. And even filtration by data source doesn't help.

Reply:
This sounds like things are working -- open_access_links should be getting all resources associated with that identifier, which might include the same book from multiple sources. Maybe `best_open_access_resource` is just reaching the wrong decision in your situation? What are the different data sources you have providing the same book? What makes one better than another?

Reply:
There is nothing wrong about data sources, somehow we just ended up with multiple data sources basically having the same resources pointing to the same URLs:
```
- Data source # 1 <- hyperlink # 1 <- resource # 1 with `http://books.com/book1.epub`
- Data source # 2 <- hyperlink # 2 <- resource # 2 with `http://books.com/book1.epub`
...
- Data source # N <- hyperlink # N <- resource # N with `http://books.com/book1.epub`
```

After we run `opds_import_monitor` for a particular collection (let's say related to data source # 2) we end up with the following state in the database (note that only resource # 2's URL has been updated):
```
- Data source # 1 <- hyperlink # 1 <- resource # 1 with `http://books.com/book1.epub`
- Data source # 2 <- hyperlink # 2 <- resource # 2 with `http://books.com/book1234567890.epub`
...
- Data source # N <- hyperlink # N <- resource # N with `http://books.com/book1.epub`
```
`LicensePool. best_open_access_resource` without filtration by a data source fetches the first resource and we end up with the wrong, not updated URL `http://books.com/book1.epub` instead of `http://books.com/book1234567890.epub`.

Reply:
Deleting the old resources will work as a one-time measure, but it sounds like this problem can happen in the normal course of events.

You were able to look at this data and know that the first resource was wrong and the second resource was correct. Is there a way to encode that knowledge into `best_open_access_resource`?  Maybe we could add a rule that if there are two indistinguishable resources, the one that was imported later should take precedence.